File: Capítulo_8/Sondagem linear/Linear probing lazy/Código em Python/hashTableLinearProbingLazy.py
# defaultToString:
def defaul_to_string(item):
    if item is None:
        return "NONE"
    elif isinstance(item, str):
        return f"{item}"
    return str(item)

# Classe ValuePair
class ValuePair:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __str__(self):
        return f"[#{self.key}: {self.value}]"

# Classe ValuePairLazy
class ValuePairLazy(ValuePair):
    def __init__(self, key, value, is_deleted = False):
        super().__init__(key, value)
        self.key = key
        self.value = value
        self.is_deleted = is_deleted

# Classe HashTableLinearProbingLazy
class HashTableLinearProbingLazy:
    def __init__(self, to_str_fn = defaul_to_string):
        self.to_str_fn = to_str_fn
        self.__table = {}
    
    # Método getter para __table
    def get_table(self):
        return self.__table

    # Método setter para __table
    def set_table(self, key, item):
        self.__table[key] = item
    
    # Método loselose_hash_code
    def loselose_hash_code(self, key):
        if isinstance(key, (int, float)):
            return key
        table_key = self.to_str_fn(key)
        hash = 0
        for i in range(len(table_key)):
            hash += ord(table_key[i])
        return hash % 37
    
    # Método hash_code
    def hash_code(self, key):
        return self.loselose_hash_code(key)
    
    # Método put
    def put(self, key, value):
        if key and value is not None:
            position = self.hash_code(key)
            table = self.get_table()
            if (table.get(position) is None or (table.get(position) is not None and table.get(position).is_deleted)):
                self.set_table(position, ValuePairLazy(key, value))
            else:
                index = position + 1
                while (table.get(index) is not None and table.get(index).is_deleted):
                    index += 1
                self.set_table(index, ValuePairLazy(key, value))
            return True
        return False
    
    # Método get
    def get(self, key):
        position = self.hash_code(key)
        table = self.get_table()
        if table.get(position) is not None:
            if table.get(position).key == key and not table.get(position).is_deleted:
                return table.get(position).value
            index = position + 1
            while table.get(index) is not None and (table.get(index).key != key or table.get(index).is_deleted):
                if table.get(index).key == key and table.get(index).is_deleted:
                    return None
                index += 1
            if table.get(index) is not None and table.get(index).key == key and not table.get(index).is_deleted:
                return table.get(index).value
        return None
    
    # Método remove
    def remove(self, key):
        position = self.hash_code(key)
        table = self.get_table()
        if table.get(position) is not None:
            if table.get(position).key == key and not table.get(position).is_deleted:
                table[position].is_deleted = True
                return True
            index = position + 1
            while table.get(index) is not None and (table.get(index). key != key or table.get(index).is_deleted):
                index += 1
            if table.get(index) is not None and table.get(index). key == key and not table.get(index).is_deleted:
                table[index].is_deleted = True
                return True
            return False
    
    # Método size
    def size(self):
        count = 0
        for value_pair in self.get_table().values():
            if value_pair and not value_pair.is_deleted:
                count += 1
        return count

    # Método is_empty
    def is_empty(self):
        return self.size() == 0
    
    # Método clear
    def clear(self):
        self.__table = {}

    # Método to_string
    def to_string(self):
        if self.is_empty():
            return ""
        keys = list(self.get_table().keys())
        obj_string = f"{{{keys[0]} => {self.get_table()[keys[0]].__str__()}}}"
        for i in range(1, len(keys)):
            obj_string = f"{obj_string}, {{{keys[i]} => {self.get_table()[keys[i]].__str__()}}}"
        return obj_string

# Criação da instância da tabela hash
hashTable = HashTableLinearProbingLazy()

# Adicionando pares chave-valor
hashTable.put('Gandalf', 'gandalf@email.com')
hashTable.put('John', 'johnsnow@email.com')
hashTable.put('Tyrion', 'tyrion@email.com')

# Recuperando valores
print(hashTable.get('Gandalf')) #Output: gandalf@email.com
print(hashTable.get('John')) # Output: johnsnow@email.com
print(hashTable.get('Tyrion')) #Output: tyrrion@email.com
print(hashTable.get('Unknown')) #Output: None

# Removendo valores
print(hashTable.remove('John')) # Output: true
print(hashTable.get('John')) # Output: None

# Exibindo o conteúdo da tabela hash
print (hashTable.to_string()) # Pode mostrar algo como {19 => [Gandalf: gandalf@email.com]}, {16 => [Tyrion: tyrion@email.com]}

for position, value_pair in hashTable.get_table().items():
    print(f"Position: {position}, Key: {value_pair.key}, Value: {value_pair.value}, Deleted: {value_pair.is_deleted}")

# Limpando a tabela hash
hashTable.clear()
print(hashTable.is_empty()) # Output: true
print(hashTable.size()) # Output: 0

File: Capítulo_8/Sondagem linear/Linear probing lazy/hashTableLinearProbingLazy.js
import { defaultToString, ValuePairLazy } from "../../util/util.js";

// Classe HashTableLinearProbingLazy
class HashTableLinearProbingLazy {
    #table;
    constructor(toStrFn = defaultToString) {
        this.toStrFn = toStrFn;
        this.#table = {};
    }

    // Método getter para #table
    _getTable() {
        return this.#table;
    }

    // Método setter para #table
    _setTable(key, item) {
        this.#table[key] = item;
    }

    // Método loseloseHashCode
    loseloseHashCode(key) {
        if (typeof key === "number") {
            return key;
        }
        const tableKey = this.toStrFn(key);
        let hash = 0;
        for (let i = 0; i < tableKey.length; i++) {
            hash += tableKey.charCodeAt(i);
        }
        return hash % 37;
    }

    // Método hashCode
    hashCode(key) {
        return this.loseloseHashCode(key);
    }

    // Método put
    put(key, value) {
        if (key != null && value != null) {
            const position = this.hashCode(key);
            if (
                this._getTable()[position] == null ||
                (this._getTable()[position] != null &&
                    this._getTable()[position].isDeleted)
            ) {
                this._setTable(position, new ValuePairLazy(key, value));
            } else {
                let index = position + 1;
                while (
                    this._getTable()[index] != null &&
                    this._getTable()[position].isDeleted
                ) {
                    index++;
                }
                this._setTable(index, new ValuePairLazy(key, value));
            }
            return true;
        }
        return false;
    }

    // Método get
    get(key) {
        const position = this.hashCode(key);
        const table = this._getTable();
        if (table[position] != null) {
            if (table[position].key === key && !table[position].isDeleted) {
                return table[position].value;
            }
            let index = position + 1;
            while (
                table[index] != null &&
                (table[index].key !== key || table[index].isDeleted)
            ) {
                if (table[index].key === key && table[index].isDeleted) {
                    return undefined;
                }
                index++;
            }
            if (
                table[index] != null &&
                table[index].key === key &&
                !table[index].isDeleted
            ) {
                return table[index].value;
            }
        }
        return undefined;
    }

    // Método remove
    remove(key) {
        const position = this.hashCode(key);
        if (this._getTable()[position] != null) {
            if (
                this._getTable()[position].key === key &&
                !this._getTable()[position].isDeleted
            ) {
                this._getTable()[position].isDeleted = true;
                return true;
            }
            let index = position + 1;
            while (
                this._getTable()[index] != null &&
                (this._getTable()[index].key !== key ||
                    this._getTable()[index].isDeleted)
            ) {
                index++;
            }
            if (
                this._getTable()[index] != null &&
                this._getTable()[index].key === key &&
                !this._getTable()[index].isDeleted
            ) {
                this._getTable()[index].isDeleted = true;
                return true;
            }
        }
        return false;
    }

    // Método isEmpty
    isEmpty() {
        return this.size() === 0;
    }

    // Método size
    size() {
        let count = 0;
        Object.values(this._getTable()).forEach((valuePair) => {
            count += valuePair.isDeleted === true ? 0 : 1;
        });
        return count;
    }

    // Método clear
    clear() {
        this.#table = {};
    }

    // Método toString
    toString() {
        if (this.isEmpty()) {
            return "";
        }
        const keys = Object.keys(this._getTable());
        let objString = `{${keys[0]} => ${this._getTable()[
            keys[0]
        ].toString()}}`;
        for (let i = 1; i < keys.length; i++) {
            objString = `${objString}, {${keys[i]} => ${this._getTable()[
                keys[i]
            ].toString()}}`;
        }
        return objString;
    }
}

// Criação da instância da tabela hash
const hashTable = new HashTableLinearProbingLazy(defaultToString);

// Adicionando pares chave-valor
hashTable.put('Gandalf', 'gandalf@email.com');
hashTable.put('John', 'johnsnow@email.com');
hashTable.put('Tyrion', 'tyrion@email.com');

// Recuperando valores
console.log(hashTable.get('Gandalf')); // Output: gandalf@email.com
console.log(hashTable.get('John')); // Output: johnsnow@email.com
console.log(hashTable.get('Tyrion')); // Output: tyrrion@email.com
console.log(hashTable.get('Unknown')); // Output: undefined

// Removendo valores
console.log(hashTable.remove('John')); // Output: true
console.log(hashTable.get('John')); // Output: undefined

// Exibindo o conteúdo da tabela hash
console.log(hashTable.toString()); 
// Pode mostrar algo como {19 => [Gandalf: gandalf@email.com]}, {16 => [Tyrion: tyrion@email.com]}

// Limpando a tabela hash
hashTable.clear();
console.log(hashTable.isEmpty()); // Output: true
console.log(hashTable.size()); // Output: 0
